### ISSUE TITLE
Probable memory leak in ArchiveIterator

### ISSUE THREAD

**Issue by kartheek7895** [webrecorder/warcio#34]
In ArchiveIterator, chunks of data(16384) are being decompressed using `decompressor=zlib.decompressobj() ` however once decompressing is done the `decompressor.flush()` is not done,  which is leaking memory when reading large files or large number of files in python 2.7. Please look into it @ikreymer

**Comment by kartheek7895** [comment]
Not sure if this is the exact obj leaking memory but memory leak is happening while trying to read a record

**Comment by ikreymer** [comment]
Thanks for the report, added a fix which calls `decompressor.flush()` on close and an `ArchiveIterator.close()`

With this, you can read a single record and wrap the call in `closing` contextmanager

```
from contextlib import closing

with open('filename.warc.gz') as fh:
     with closing(ArchiveIterator(fh)) as ai:
         record = ai.next()
         ...
```

Let me know if this works for you. The fix is on the develop branch and will be part of the next release.

**Comment by ikreymer** [comment]
Can you provide more info about your use case? Are you opening a file, reading one record, and then closing after use? I haven't seen any indication of memory leaks with warcio..

**Comment by kartheek7895** [comment]
Yes, I will be having a bunch of warc files on a local machine and based on requirement the script will have to extract the required record and save it to a file. PFB, more details about the use case and code. Let me know if you need any more details.
```
Filename: justWarcTester.py

Line #    Mem usage    Increment   Line Contents
================================================
    82     21.8 MiB     21.8 MiB   	@profile
    83                             	def fetchFromWarcFile(self,fileName,offset):
    84     21.8 MiB      0.0 MiB   			try:
    86     21.8 MiB      0.0 MiB   				f = open(fileName, "rb")
    87     21.8 MiB      0.0 MiB   				f.seek(int(offset))
    89     21.8 MiB      0.0 MiB   				with closing(warcio.ArchiveIterator(f)) as it:
    90     22.1 MiB      0.3 MiB   					record = next(it)                          
    95     22.1 MiB      0.0 MiB   					url = record.rec_headers.get_header('WARC-Target-URI')
   104     22.1 MiB      0.0 MiB   					urlh = hashlib.sha1(url.encode()).hexdigest()
   120     22.1 MiB      0.0 MiB   					with open ("./test/"+urlh+".jpg","wb") as fileWriter:
   121     22.1 MiB      0.0 MiB   						fileWriter.write(record.raw_stream.read())
   123     22.1 MiB      0.0 MiB   					f.close()
   125                             			except Exception as e:
   126                             				print (e)
```

Driver code:
```
if __name__ =="__main__":
        data = json.load(open("testData.json","r"))
        for i in range(0,1000):
                for ent in data:
                        add = ent["file_name"]
                        offset = ent["offset"]
                        obj = PushWarcClass()
                        obj.fetchFromWarcFile(fileName=add,offset=offset)
```

However the `0.3 MiB` that `next(it)` is taking is not being released by the process, in the next iteration which is slowly adding up.

**Comment by kartheek7895** [comment]
Any updates on this?

**Comment by ikreymer** [comment]
Sorry, I haven't had time to look at this too much further.. If it's working in python 3.5+, I think that may be good enough to release. There may be other issues in 2.7 that may have been fixed in python 3. If you have time to debug this and identify a solution, will be happy to accept a PR.

**Comment by ikreymer** [comment]
Released in warcio 1.5.2!

**Comment by prnake** [comment]
I believe this issue still exists. In one situation, when the error at [here](https://github.com/webrecorder/warcio/blob/e0eee8eca035edd3c4b5d4884449ae3f5b0abfd4/warcio/archiveiterator.py#L112) is triggered, memory flush will not be automatically called.
